torch/distributed/pipelining/_backward.py

```diff
@@ -1,12 +1,234 @@
 # mypy: allow-untyped-defs
 # Copyright (c) Meta Platforms, Inc. and affiliates
-from typing import List, Optional
+import collections
+import logging
+import weakref
+from typing import Any, cast, Deque, Dict, Iterator, List, Optional, Set, Tuple, Union
 
 import torch
+from torch.autograd.graph import GradientEdge, Node
+from torch.nn import Parameter
 
 from ._debug import map_debug_info
 
 
+logger = logging.getLogger(__name__)
+
+
+def _get_grad_fn_or_grad_acc(t: torch.Tensor) -> Union[Node, None]:
+    if t.requires_grad and t.grad_fn is None:
+        # if no grad function (leaf tensors) we use view
+        viewed_t = t.view_as(t)
+        grad_fn = viewed_t.grad_fn
+        if grad_fn is not None:
+            return grad_fn.next_functions[0][0]
+        else:
+            raise RuntimeError("grad_fn after adding view is None")
+    else:
+        return t.grad_fn
+
+
+def reverse_closure(
+    roots: List[Node], target_nodes: Set[Node]
+) -> Tuple[Set[Node], Set[Node]]:
+    # Recurse until we reach a target node
+    closure: Set[Node] = set()
+    actual_target_nodes = set()
+    q: Deque[Node] = collections.deque()
+    for node in roots:
+        if node is not None and node not in closure:
+            closure.add(node)
+            q.append(node)
+    while q:
+        node = q.popleft()
+        metadata = cast(Dict[str, List], node.metadata)
```

Comment:
where does node.metadata actually come from? autograd gives it to us and we're just formally building a python weakref graph out of it? or we added it somewhere else?  would be good to comment the top of the function wth the overall semantics and why we want the reverse_closure

Reply:
ok we add it below. and i guess 'reverse closure' refers to a closure over a reversed graph?

Reply:
> where does node.metadata actually come from? autograd gives it to us and we're just formally building a python weakref graph out of it? or we added it somewhere else?

autograd gives it to us

Reply:
Will add comments. autograd give it to us and we use it to traverse the graph. I need to do the casting because `Node` in graph.py has it defined as a method (https://github.com/pytorch/pytorch/blob/main/torch/autograd/graph.py#L77-L80), whereas it's used as just a dict, maybe that should be updated?